### PR TITLE
test: fix non-deterministic backfill test

### PR DESCRIPTION
**Pull request by cmwaters**
Closes: https://github.com/tendermint/tendermint/issues/6644

This removes the use of math/rand so that the tests should pass deterministically.

I also added a check for leaky go routines which uncovered a situation where go routines would continue after the function was finished.

I ran:
```
go test -run TestReactor_Backfill -count 20
``` 
locally and it seemed to work

### CODE REVIEW

File: internal/statesync/reactor_test.go [hunk 1]
```diff
@@ -3,12 +3,12 @@ package statesync
 import (
 	"context"
 	"fmt"
-	"math/rand"
 	"sync"
 	"testing"
 	"time"
 
 	// "github.com/fortytw2/leaktest"
```
Comment (marbar3778) [inline internal/statesync/reactor_test.go:10]:
```suggestion
```

**Review by marbar3778 — APPROVED** [review]
utACK

File: internal/statesync/reactor.go [hunk 2]
```diff
@@ -310,12 +310,17 @@ func (r *Reactor) backfill(
 	// waiting on blocks. If it takes 4s to retrieve a block and 1s to verify
 	// it, then steady state involves four workers.
 	for i := 0; i < int(r.cfg.Fetchers); i++ {
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer cancel()
```
Comment (williambanfield) [inline internal/statesync/reactor.go:314]: nice catch.

**Review by williambanfield — COMMENTED** [review]
do you know why the tests would fail when it was non-deterministic?

**Comment by cmwaters** [comment]
> do you know why the tests would fail when it was non-deterministic?

Yeah basically I run three different backfills with varying levels of faults from a node's peers. But the amount of faults from a peer was random so occasionally the peers would fault more often then the threshold I had set and thus occasionally it would fail

**Pull request merged**